sklearn/metrics/classification.py

```diff
@@ -1596,37 +1603,55 @@ def log_loss(y_true, y_pred, eps=1e-15, normalize=True, sample_weight=None):
     -----
     The logarithm used is the natural logarithm (base-e).
     """
+    y_pred = check_array(y_pred, ensure_2d=False)
+    check_consistent_length(y_pred, y_true)
+
     lb = LabelBinarizer()
-    T = lb.fit_transform(y_true)
-    if T.shape[1] == 1:
-        T = np.append(1 - T, T, axis=1)
 
-    y_pred = check_array(y_pred, ensure_2d=False)
+    if labels is not None:
+        lb.fit(labels)
+    else:
+        lb.fit(y_true)
+
+    if len(lb.classes_) == 1:
+        if labels is None:
+            raise ValueError('y_true has only one label. Please provide '
+                             'the true labels explicitly through the labels '
+                             'argument.')
+        else:
+            raise ValueError('labels should have more than one unique label.')
+
+    transformed_labels = lb.transform(y_true)
+
+    if transformed_labels.shape[1] == 1:
+        transformed_labels = np.append(1 - transformed_labels,
+                                       transformed_labels, axis=1)
+
     # Clipping
-    Y = np.clip(y_pred, eps, 1 - eps)
+    y_pred = np.clip(y_pred, eps, 1 - eps)
 
-    # This happens in cases when elements in y_pred have type "str".
-    if not isinstance(Y, np.ndarray):
-        raise ValueError("y_pred should be an array of floats.")
+    # if not isinstance(y_pred, np.ndarray):
+    #     raise ValueError("y_pred should be an array of floats.")
 
     # If y_pred is of single dimension, assume y_true to be binary
     # and then check.
-    if Y.ndim == 1:
-        Y = Y[:, np.newaxis]
-    if Y.shape[1] == 1:
-        Y = np.append(1 - Y, Y, axis=1)
+    if y_pred.ndim == 1:
+        y_pred = y_pred[:, np.newaxis]
+    if y_pred.shape[1] == 1:
+        y_pred = np.append(1 - y_pred, y_pred, axis=1)
 
     # Check if dimensions are consistent.
-    check_consistent_length(T, Y)
-    T = check_array(T)
-    Y = check_array(Y)
-    if T.shape[1] != Y.shape[1]:
+    transformed_labels = check_array(transformed_labels)
+    if transformed_labels.shape[1] != y_pred.shape[1]:
         raise ValueError("y_true and y_pred have different number of classes "
-                         "%d, %d" % (T.shape[1], Y.shape[1]))
+                         "%d, %d. Please provide the true labels explicitly "
```

Comment:
I thought I had commented here :-( . 

When you provide labels, it is assumed that these labels are from the entire training data, and your y_score matrix is got after fitting your model on the entire training data. So `len(labels) == y_score.shape[1]` which is also equal to `len(lb.classes_)` since you fit your binarizer with `labels`.

If you do not provide labels, it is assumed `y_true` has all possible labels. Hence still `len(np.unique(y_true)) == y_pred.shape[1]` which is again equal to `len(lb.classes_)` since now you fit the binarizer with `y_true`.

> what do you mean "appended 1-y_pred for that case above"? 

I meant that for the `(n_samples,)` (binary) case you have explicitly converted to `(n_samples, 2)` above right?